README.md
# CMS-iOS

This is the iOS version on the CMS app, written in Swift. The app utilises the API endpoints provided by Moodle from the server hosted on the [CMS Website](https://td.bits-hyderabad.ac.in/moodle/).

<a href="https://apps.apple.com/in/app/cms-bphc/id1489946522">
  <img height="40" src="https://user-images.githubusercontent.com/7317008/43209852-4ca39622-904b-11e8-8ce1-cdc3aee76ae9.png">
</a>

[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](https://github.com/dwyl/esta/issues)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

## Features

- Token based as well as **Google** login.
- Offline access to course contents.
- Enrollment to new courses.
- Unenrollment from existing courses.
- Local storage of attachments for external access from Files app.
- Push Notifications

## Installation

The deployment target for the app is iOS 11.0, so if your device is not on that or a newer version, you're out of luck `¯\_(ツ)_/¯` 

You can download the latest release of the app from the App Store [here](https://apps.apple.com/in/app/cms-bphc/id1489946522).

If you want to manually install the app from this repository, first clone it. Navigate to the project directory and then install the pods by executing `pod install`. Once pods have been installed successfully, open the file named `CMS-iOS.xcworkspace` and then change the development profile to your profile and then run it after selecting your device from the drop-down next to the run button.

## Contributing

You can contribute by submitting an issue, or by picking an existing one from the list and sending a PR. Ensure that the branch underlying the PR is up-to-date with latest changes.

All changes must be made on top of a new branch forked off the `development` branch. The `master` branch is only updated when a new release is made.

Commit messages should follow common guidelines, such as the ones mentioned [here](https://chris.beams.io/posts/git-commit/), whenever possible.

Mention keywords such as "Fixes" or "Closes" in commit messages, followed by the issue number, to automatically close corresponding issues. [(List of keywords)](https://help.github.com/articles/closing-issues-using-keywords/)

The Moodle endpoints used by this app have been collected in a [Postman](https://getpostman.com") [here](https://www.getpostman.com/collections/207390c4552febe4e393). You may import this collection into Postman directly using the link. 
## License

The app is under the MIT License.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.3.2] - 2022-04-16

### Fixed
- Dashboard crashes due to Realm object invalidation
- Long course title label on Dashboard cells
- Crashes on clicking a notification


### Removed
- Login view as initial view and replaced it with the Dashboard view

## [1.3.1] - 2021-10-24

### Fixed
- Crash on long-pressing a module or discussion on Dashboard during search
- HTML2String function, that caused issues on Site News page
- Custom tab bar with black background in Light Mode

## [1.3] - 2021-08-22

### Fixed
- Fixed unenrollment process, removed need for sign in through web view using a private token obtained during Google Login

### Removed
- Semester label in Dashboard cell and option to show/hide it

## [1.2] - 2021-01-17

### Added
- Support for push notifications
- Redirection to module/discussion when notifications are clicked
- Unenrolling feature
- Support for adding image of timetable
- Quick loading of cached discussions and site news
- Refresh control for discussions
- View models for cells of Dashboard courses, discussions and site news

### Fixed
- Random numbers showing up in Dashboard course cell badges
- Some course discussions being marked unread after reloading site news due to matching ids


## [1.1.1] - 2021-01-12
### Added
- Support for CMS on AWS by adding new base URL.

### Removed
- Notifications using background fetch.

## [1.1] - 2020-08-23
### Added
- Ability to search for modules and courses on Dashboard itself
- Redesigned Extras Tab
- Semester Label in Dashboard cell
- Option to hide semester label
- System spotlight search indexing for courses
- Realm Migration Block for future use

### Fixed
- Colors of cells in Light Mode on Dashboard to be more streamlined with the rest of the UI
- Crash when the app is opened using URL Schemes as it would previously work only with Login With Google
- Small titles for courses
- Incorrect Regex parsing for some courses
- HTML entities showing up in action sheet
- Infinitely loading refresh control on course contents when offline
- Bug where course contents would show up in incorrect order while offline
- Color Allotment blocking main thread and leading to lag on Dashboard
- Bug where user would sometimes by spammed with notifications
- Crash on opening the app using URL schemes

### Removed
- Inconsistent and buggy table view animation

## [1.0.1] - 2020-02-19
### Added
- Support for viewing forum module description
### Fixed
- Crash on downloading modules on iOS 13
- HTML entities showing up on course content page


CMS-iOS/AppDelegate.swift
//
//  AppDelegate.swift
//  CMS-iOS
//
//  Created by Hridik Punukollu on 09/08/19.
//  Copyright © 2019 Crux BPHC. All rights reserved.
//

import UIKit
import Realm
import RealmSwift
import IQKeyboardManagerSwift
import UserNotifications
import SDDownloadManager
import SafariServices
import CoreSpotlight
import SwiftKeychainWrapper
import Alamofire
import SVProgressHUD

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterDelegate {
    
    var window: UIWindow?
    
    let notificationCenter = UNUserNotificationCenter.current()
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
//        UIApplication.shared.registerForRemoteNotifications()
//        application.setMinimumBackgroundFetchInterval(900)
        
        // Code for realm migration, update this when realm schema is changed
        
        /*
         Migration 2: Added update flag on courses
         */

        let config = Realm.Configuration(
            schemaVersion: 2, // version to change schema to
            // Set the block which will be called automatically when opening a Realm with
            // a schema version lower than the one set above
            migrationBlock: { migration, oldSchemaVersion in
                if (oldSchemaVersion < 2) {
                    // nothing to be done here, only property was added
                }
            })
        
        // Tell Realm to use this new configuration object for the default Realm
        Realm.Configuration.defaultConfiguration = config
        
        let options : UNAuthorizationOptions = [.alert, .sound, .badge]
        notificationCenter.requestAuthorization(options: options) { (didAllow, error) in
            guard didAllow else {return}
//            BackgroundFetch().setCategories()
//            if !didAllow {
//                print("The user denied notification permission.")
//            } else if didAllow {
//                BackgroundFetch().setCategories()
//            }
        }
        
        // Set SVProgressHUD theme
        if #available(iOS 13.0, *) {
            if UITraitCollection.current.userInterfaceStyle == .dark{
                SVProgressHUD.setDefaultStyle(.dark)
            } else {
                SVProgressHUD.setDefaultStyle(.light)
            }
        } else {
            SVProgressHUD.setDefaultStyle(.dark)
        }
        
        Alamofire.SessionManager.default.delegate.taskWillPerformHTTPRedirection = nil
        
        
        UIApplication.shared.registerForRemoteNotifications()
        IQKeyboardManager.shared.enable = true
        
//        if let userInfo = launchOptions?[.remoteNotification] as? [AnyHashable: Any] {
////            manageNotificationPayload(userInfo: userInfo)
//        }
        return true
    }
    
    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and invalidate graphics rendering callbacks. Games should use this method to pause the game.
    }
    
    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }
    
    
    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
        UIApplication.shared.applicationIconBadgeNumber = 0
    }
    
    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }
    
    func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey : Any] = [:]) -> Bool {
        
        guard let message = url.host else { return true }
        guard let tabVC = self.window?.rootViewController as? BubbleTabBarController else { return false }
        tabVC.selectedIndex = 0
        guard let navVC = tabVC.viewControllers?.first as? UINavigationController else { return false }
        guard let dashboardVC = (navVC.viewControllers.first as? DashboardViewController) else { return false }
        guard let loginVC = dashboardVC.loginViewController else { return false }
        loginVC.loginWithGoogle(input: message)
        loginVC.safariVC.dismiss(animated: true)
        
        return true
    }
    
    //    func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any], fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
    //        <#code#>
    //    }
    
    func application(_ application: UIApplication, handleEventsForBackgroundURLSession identifier: String, completionHandler: @escaping () -> Void) {
        SDDownloadManager.shared.backgroundCompletionHandler = completionHandler
    }
    
//    func application(_ application: UIApplication, performFetchWithCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
//        if Reachability.isConnectedToNetwork() {
//            let bkgObj = BackgroundFetch()
//            bkgObj.downloadModules { newModulesFound in
//                let realm = try! Realm()
//                let discussionModules = realm.objects(Module.self).filter("modname = %@", "forum")
//                bkgObj.downloadDiscussions(discussionModules: discussionModules) { (newDiscussionsFound) in
//                    completionHandler(newDiscussionsFound || newModulesFound ? .newData : .noData)
//                    NSLog(newDiscussionsFound || newModulesFound ? "found new data" : "no new data found")
//                }
//            }
//        } else {
//            completionHandler(.failed)
//            print("failed to background fetch")
//        }
//    }
    
    func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
        
        let token = deviceToken.map { String(format: "%02.2hhx", $0) }.joined()
        print("device token = \(token)")
        KeychainWrapper.standard.set(token, forKey: "deviceToken")
//        NotificationManager.shared.registerDevice(deviceToken: token) {
            
//        }
    }
    
    func application(_ application: UIApplication, didFailToRegisterForRemoteNotificationsWithError error: Error) {
        
        
    }
    
    func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
            if response.actionIdentifier == "Mark as Read" {

                //Mark as read function
            } else if response.actionIdentifier == "Open" {
    //            Open file
            }
        }
    
    func userNotificationCenter(_ center: UNUserNotificationCenter,
                                willPresent notification: UNNotification,
                                withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
        // Print message ID.
        UNUserNotificationCenter.current().delegate = self
        // Print full message.
    }
    // handle opening with spotlight search
    func application(_ application: UIApplication, continue userActivity: NSUserActivity, restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void) -> Bool {
        if userActivity.activityType == CSSearchableItemActionType {
            if let uniqueIdentifier = userActivity.userInfo?[CSSearchableItemActivityIdentifier] as? String {
                guard let typeId = getTypeAndId(string: uniqueIdentifier) else { return true }
                self.manageRedirection(redirectType: typeId.keys.first!, id: typeId.values.first!)
                
            }
        }
        
        return true
    }
    
    private func getTypeAndId(string: String) -> [String: Int]? {
        let comps = string.components(separatedBy: "=")
        let first = comps.first
        let last = Int(comps.last!)
        if first != nil && last != nil {
            return [first!: last!]
        }
        return nil
    }
    
}

// Managing Notifications
extension AppDelegate {
    
    func manageRedirection(redirectType: String, id: Int) {
        let keyWindow = UIApplication.shared.windows.filter { $0.isKeyWindow }.first
        
        if var topController = keyWindow?.rootViewController {
            while let presentedViewController = topController.presentedViewController {
                topController = presentedViewController
            }
            
            // topController should now be your topmost view controller
            if let bubbleTabVC = topController as? BubbleTabBarController {
                // app is already open
                bubbleTabVC.selectedIndex = 0
                guard let navigationVC = bubbleTabVC.viewControllers?.first as? UINavigationController else { return }
                let realm = try! Realm()
                switch redirectType {
                case "course":
                    guard let course = realm.objects(Course.self).filter("courseid = %@", id).first else { return }
                    let storyboard = UIStoryboard(name: "Main", bundle: nil)
                    let courseContentsVC = storyboard.instantiateViewController(withIdentifier: "Course Contents") as! CourseDetailsViewController
                    courseContentsVC.currentCourse = course
                    navigationVC.pushViewController(courseContentsVC, animated: true)
                    break
                case "module":
                    // get the module and push it
                    guard let module = realm.objects(Module.self).filter("id = %@", id).first else { return }
                    let storyboard = UIStoryboard(name: "Main", bundle: nil)
                    let moduleVC = storyboard.instantiateViewController(withIdentifier: "Module View Controller") as! ModuleViewController
                    moduleVC.selectedModule = module
                    navigationVC.pushViewController(moduleVC, animated: true)
                    break
                case "discussion":
                    // get the discussion object
                    guard let discussion = realm.objects(Discussion.self).filter("id = %@", id).first else { return }
                    let storyboard = UIStoryboard(name: "Main", bundle: nil)
                    let discussionVC = storyboard.instantiateViewController(withIdentifier: "Discussions") as! DiscussionViewController
                    discussionVC.selectedDiscussion = discussion
                    navigationVC.pushViewController(discussionVC, animated: true)
                    break
                case "section":
                    guard let course = realm.objects(Course.self).filter("courseid = %@", id).first else { return }
                    let storyboard = UIStoryboard(name: "Main", bundle: nil)
                    let courseContentsVC = storyboard.instantiateViewController(withIdentifier: "Course Contents") as! CourseDetailsViewController
                    courseContentsVC.currentCourse = course
                    navigationVC.pushViewController(courseContentsVC, animated: true)
                    break
                default:
                    break
                }
                return
                
            } else {
                // app is not open, proceed with login vc
                guard let loginVC = topController as? LoginViewController else { return }
                loginVC.redirectTo = [redirectType: id]
                
            }
            
        }
    }
    
    func manageNotificationPayload(userInfo: [AnyHashable: Any]) {
        let keyWindow = UIApplication.shared.windows.filter { $0.isKeyWindow }.first
        
        guard let topController = keyWindow?.rootViewController else { return }
        topController.showLoadingIndicator(message: "Loading...")
        guard let respDict = userInfo["aps"] as? NSDictionary else { return }
        if let courseIdString = (respDict["data"] as? NSDictionary)?["courseid"] as? String {
            guard let courseId = Int(courseIdString) else { return }
            print("Received notification with  course id:", courseId)
            guard let contextURL = (respDict["data"] as? NSDictionary)?["contexturl"] as? String else { return }
            guard let name = (respDict["data"] as? NSDictionary)?["name"] as? String else { return }
            var notificationType = ""
            if courseId == 1 {
                notificationType = "siteNews"
            } else if contextURL.contains("discuss.php") {
                notificationType = "discussion"
            } else if contextURL.contains("/mod/") {
                notificationType = "module"
            } else if contextURL.contains("/course/") {
                notificationType = "section"
            }
            switch notificationType {
            case "discussion":
                guard let discussionIdString = Regex.match(pattern: "#p[0-9]+", text: contextURL).first?.replacingOccurrences(of: "#p", with: "") else { return }
                guard let discussionId = Int(discussionIdString) else { return }
                DashboardDataManager.shared.getAndStoreDiscussions(forCourse: courseId) {
                    Alamofire.SessionManager.default.session.getAllTasks { (tasks) in
                        tasks.forEach{ $0.cancel() }
                    }
                    DispatchQueue.main.async {
                        topController.dismissLoadingIndicator()
                        self.manageRedirection(redirectType: notificationType, id: discussionId)
                    }
                }
                break
            case "module":
                guard let moduleIdString = Regex.match(pattern: "id=[0-9]+", text: contextURL).first?.replacingOccurrences(of: "id=", with: "") else { return }
                guard let moduleId = Int(moduleIdString) else { return }
                DashboardDataManager.shared.getAndStoreModules(forCourse: courseId) {
                    DispatchQueue.main.async {
                        topController.dismissLoadingIndicator()
                        self.manageRedirection(redirectType: notificationType, id: moduleId)
                    }
                }
                break
            case "section":
                topController.dismissLoadingIndicator()
                self.manageRedirection(redirectType: notificationType, id: courseId)
                break
            case "siteNews":
                topController.dismissLoadingIndicator()
                if let bubbleTabVC = topController as? BubbleTabBarController {
                    bubbleTabVC.selectedIndex = 2
                }
                break
            default:
                topController.dismissLoadingIndicator()
                break
            }
        }
    }
    
    
    func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any], fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
        print(userInfo)
        
        if application.applicationState == .active {
            
            // app is currently active
            
        } else if application.applicationState == .background {
            
            //app is in background, if content-available key of your notification is set to 1, poll to your backend to retrieve data and update your interface here
            
        } else if application.applicationState == .inactive {
            // user tapped on the notification, do all the stuff here
            manageNotificationPayload(userInfo: userInfo)
            
        }
        
        
    }
        
}
